server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -1910,13 +1919,31 @@ final RefreshResult refresh(String source, SearcherScope scope, boolean block) t
 
     @Override
     public void writeIndexingBuffer() throws EngineException {
+        // If we're already halfway through the flush thresholds, then we do a flush. This will save us from writing segments twice
+        // independently in a short period of time, once to reclaim IndexWriter buffer memory and then to reclaim the translog.
+        final long flushThresholdSizeInBytes = Math.max(
+            Translog.DEFAULT_HEADER_SIZE_IN_BYTES + 1,
+            config().getIndexSettings().getFlushThresholdSize(totalDiskSpace).getBytes() / 2
+        );
```

Comment:
Agreed, I'll improve the comment to make it more explicit. Hopefully this becomes a bit less of a concern when we write a single segment to reclaim memory instead of writing all pending segments.